primitives/runtime-interface/proc-macro/src/runtime_interface/host_function_interface.rs

```diff
@@ -194,47 +209,162 @@ fn generate_host_function_implementation(
-	let into_preallocated_ffi_value = generate_into_preallocated_ffi_value(&method.sig)?;
-	let convert_return_value = generate_return_value_into_wasm_value(&method.sig);
 
-	Ok(quote! {
-		{
-			struct #struct_name;
+	let fn_name = create_function_ident_with_version(&method.sig.ident, version);
+	let ref_and_mut = get_function_argument_types_ref_and_mut(&method.sig);
+
+	let mut ffi_args_prototype = Vec::new();
+	let mut ffi_names = Vec::new();
+	let mut host_names_with_ref = Vec::new();
+	let mut copy_data_into_ref_mut_args = Vec::new();
+	let mut convert_args_dynamic_ffi_to_static_ffi = Vec::new();
+	let mut convert_args_static_ffi_to_host = Vec::new();
+	for ((host_name, host_ty), ref_and_mut) in
+		get_function_argument_names_and_types_without_ref(&method.sig).zip(ref_and_mut)
+	{
+		let host_name_ident = match *host_name {
+			Pat::Ident(ref pat_ident) => pat_ident.ident.clone(),
+			_ => unreachable!(),
+		};
+
+		let ffi_name = generate_ffi_value_var_name(&host_name)?;
+		let ffi_ty = quote! { <#host_ty as #crate_::RIType>::FFIType };
+		ffi_args_prototype.push(quote! { #ffi_name: #ffi_ty });
+		ffi_names.push(quote! { #ffi_name });
+
+		let convert_arg_error = format!(
+			"could not marshal the '{}' argument through the WASM FFI boundary while executing '{}' from interface '{}'",
+			host_name_ident, method.sig.ident, trait_name
+		);
+		convert_args_static_ffi_to_host.push(quote! {
+			let mut #host_name = <#host_ty as #crate_::host::FromFFIValue>::from_ffi_value(__function_context__, #ffi_name)
+				.map_err(|err| format!("{}: {}", err, #convert_arg_error))?;
+		});
 
-			impl #crate_::sp_wasm_interface::Function for #struct_name {
-				fn name(&self) -> &str {
-					#name
-				}
+		let ref_and_mut_tokens =
+			ref_and_mut.map(|(token_ref, token_mut)| quote!(#token_ref #token_mut));
 
-				fn signature(&self) -> #crate_::sp_wasm_interface::Signature {
-					#signature
-				}
+		host_names_with_ref.push(quote! { #ref_and_mut_tokens #host_name });
 
-				fn execute(
-					&self,
-					__function_context__: &mut dyn #crate_::sp_wasm_interface::FunctionContext,
-					args: &mut dyn Iterator<Item = #crate_::sp_wasm_interface::Value>,
-				) -> std::result::Result<Option<#crate_::sp_wasm_interface::Value>, String> {
-					#( #wasm_to_ffi_values )*
-					#( #ffi_to_host_values )*
-					#host_function_call
-					#into_preallocated_ffi_value
-					#convert_return_value
-				}
+		if ref_and_mut.map(|(_, token_mut)| token_mut.is_some()).unwrap_or(false) {
+			copy_data_into_ref_mut_args.push(quote! {
+				<#host_ty as #crate_::host::IntoPreallocatedFFIValue>::into_preallocated_ffi_value(
+					#host_name,
+					__function_context__,
+					#ffi_name,
+				)?;
+			});
+		}
+
+		let arg_count_mismatch_error = format!(
+			"number of arguments given to '{}' from interface '{}' does not match the expected number of arguments!",
+			method.sig.ident, trait_name
+		);
```

Comment:
```suggestion
		let arg_count_mismatch_error = format!(
			"number of arguments given to '{}' from interface '{}' does not match the expected number of arguments!",
			method.sig.ident, 
			trait_name,
		);
```

Okay, here we check the signature :D

Reply:
We could also include the missing `argument` name.